src/test/java/io/confluent/examples/streams/TableRekeyingTest.java

```diff
@@ -0,0 +1,138 @@
+  private static final String outputTopic1 = "outputTopic1";
+  private static final String outputTopic2 = "outputTopic2";
+
+  @Test
+  public void shouldRekeyTheTable() {
+    final List<KeyValue<Integer, String>> inputRecords = Arrays.asList(
+        new KeyValue<>(123, "kafka"),
+        new KeyValue<>(123, "stream"),
+        new KeyValue<>(456, "stream"),
+        new KeyValue<>(456, "table"),
+        new KeyValue<>(456, "duality")
+    );
+
+    final Map<String, Integer> expectedOutput = new HashMap<>();
+    expectedOutput.put("kafka", 123);
+    expectedOutput.put("stream", 456); // Note how, for key "stream", the value is 456 and not 123.
+    expectedOutput.put("table", 456);
+    expectedOutput.put("duality", 456);
+
+    // Step 1: Create the topology and its configuration
+    final StreamsBuilder builder = createTopology();
+    final Properties streamsConfiguration = createTopologyConfiguration();
+
+    try (final TopologyTestDriver topologyTestDriver = new TopologyTestDriver(builder.build(), streamsConfiguration)) {
+      // Step 2: Write the input
+      IntegrationTestUtils.produceKeyValuesSynchronously(
+          inputTopic, inputRecords, topologyTestDriver, new IntegerSerializer(), new StringSerializer());
+
+      // Step 3: Validate the output
+
+      // Variant 1 (see explanation further down below)
+      final Map<String, Integer> actualOutput1 = IntegrationTestUtils.drainTableOutput(
+          outputTopic1, topologyTestDriver, new StringDeserializer(), new IntegerDeserializer());
+      assertThat(actualOutput1).hasSameSizeAs(expectedOutput);
+      assertThat(actualOutput1).containsAllEntriesOf(expectedOutput);
+
+      // Variant 2 (see explanation further down below)
+      final Map<String, Integer> actualOutput2 = IntegrationTestUtils.drainTableOutput(
+          outputTopic2, topologyTestDriver, new StringDeserializer(), new IntegerDeserializer());
+      assertThat(actualOutput2).hasSameSizeAs(expectedOutput);
+      assertThat(actualOutput2).containsAllEntriesOf(expectedOutput);
+    }
+  }
+
+  private StreamsBuilder createTopology() {
+    final StreamsBuilder builder = new StreamsBuilder();
+    final KTable<Integer, String> table = builder.table(inputTopic, Consumed.with(Serdes.Integer(), Serdes.String()));
+
+    // Variant 1 (https://docs.confluent.io/current/streams/faq.html#option-1-write-kstream-to-ak-read-back-as-ktable)
+    // Here, we re-key the KTable, write the results to a new topic, and then re-read that topic into a new KTable.
+    table
+        .toStream()
+        .map((key, value) -> KeyValue.pair(value, key))
+        .to(outputTopic1, Produced.with(Serdes.String(), Serdes.Integer()));
+    final KTable<String, Integer> rekeyedTable1 =
+        builder.table(outputTopic1, Consumed.with(Serdes.String(), Serdes.Integer()));
+
+    // Variant 2 (https://docs.confluent.io/current/streams/faq.html#option-2-perform-a-dummy-aggregation)
```

Comment:
This one doesn't work, actually, because reduce bypasses incoming null values (tombstones), instead of passing them into the function, so the resulting table will have "zombie" records that should have been deleted.

Reply:
should be possible to exercise this by sending an input record and then sending a tombstone for the same key.

Reply:
Hmm, there's actually more wrong this example -- even Variant 1 doesn't work. I'll remove this example from the PR and will follow-up later.

Reply:
Yeah, upon reflection, this isn't really just "rekeying" a table. It's more like a 1:many query, which would need an approach similar to https://github.com/apache/kafka/pull/5527 to be done correctly.

To extract the problem, say you have a table:
```
Table1:
1 -> A
2 -> A
```

And you want to invert the k/v relationship. Table1 isn't a bijection, so when we interpret "A" as a key, it actually has 2 values, not one, so the proper representation of the rekeyed table is:

```
Table2:
A -> {1,2}
```

Then, when we run the following updates on Table1, we should still get the right result for Table2:
```
put(2,B);

Table1:
1 -> A
2 -> B

implicit:( subtract(A, 2), add(B,2) )

Table2:
A -> {1}
B -> {2}
```
```
put(3,A);

Table1:
1 -> A
2 -> B
3 -> A

implicit:( add(A,3) )

Table2:
A -> {1,3}
B -> {2}
```

Does this seem right?

Reply:
Oh, yeah, I forgot about tombstones, but that should also work. Picking up from the last state:
```
delete(1);

Table1:
2 -> B
3 -> A

implicit:( subtract(A,1) )

Table2:
A -> {3}
B -> {2}
```

Reply:
The real brain twister is if you think of inverting Table2. Theoretically, you should get Table1 again, but there's no enforced uniqueness guarantee. Still, you have a priori knowledge that Table1:Table2 is 1:many, and therefore Table2:Table1 is many:1.

The question is whether you could naively implement the aggregation just to take every update as the aggregation result, with the following aggregation algorithm:
```
initializer: () -> null
subtractor: (key, value, agg) -> null
adder: (key, value, agg) -> value
```

Not sure if this results in a sound re-inversion. It might be interesting to test it out.